### PR TITLE
@vpnmanager script not running

### PR DESCRIPTION
**Pull request by koechkevin**
## Description

VPN manager script is not running because dist is not included in build and app.json is not in the root directory of docker build.
This PR fixes that.

Fixes # (issue)

## Type of change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation

### CODE REVIEW

**Comment by kilemensi** [comment]
_Heita_ @koechkevin ... we seem to be copying node_modules (and other test packages like jest and playwrite) even though the application uses output: standalone. Can you give context as to why?

**Comment by koechkevin** [comment]
> _Heita_ @koechkevin ... we seem to be copying node_modules (and other test packages like jest and playwrite) even though the application uses output: standalone. Can you give context as to why?

@kilemensi the scripts are running independent of next server forcing us to include node_modules.

For `other test packages like jest and playwrite`, pnpm install step was failing forcing me to include.

**Comment by kilemensi** [comment]
>@kilemensi the scripts are running independent of next server forcing us to include node_modules.
>

Seems like a heavy penalty to pay for that @koechkevin ... can't we move the script into Next.js e.g. API endpoint and just call that end point periodically?

**Comment by koechkevin** [comment]
> >@kilemensi the scripts are running independent of next server forcing us to include node_modules.
> 
> >
> 
> 
> 
> Seems like a heavy penalty to pay for that @koechkevin ... can't we move the script into Next.js e.g. API endpoint and just call that endpoint periodically?

Sure @kilemensi

**Comment by koechkevin** [comment]
@CodeForAfrica/tech

**Review by kelvinkipruto — APPROVED** [review]

**Review by m453h — APPROVED** [review]

File: .github/workflows/vpnmanager-deploy-dev.yml [hunk 1]
```diff
@@ -57,6 +57,7 @@ jobs:
             SENTRY_ENVIRONMENT=${{ env.SENTRY_ENVIRONMENT }}
             SENTRY_ORG=${{ secrets.SENTRY_ORG }}
             SENTRY_PROJECT=${{ secrets.VPNMANAGER_SENTRY_PROJECT }}
+            SECRET_TOKEN=${{ secrets.VPNMANAGER_SECRET_TOKEN }}
```
Comment (kilemensi) [inline .github/workflows/vpnmanager-deploy-dev.yml:60]: Secret for what? e.g. VPNMANAGER_SECRET_TOKEN, VPNMANAGER_API_SECRET/KEY, etc.

File: Dockerfile [hunk 2]
```diff
@@ -599,9 +601,11 @@ COPY --from=vpnmanager-builder --chown=nextjs:nodejs /workspace/apps/vpnmanager/
 # https://nextjs.org/docs/advanced-features/output-file-tracing
 COPY --from=vpnmanager-builder --chown=nextjs:nodejs /workspace/apps/vpnmanager/.next/standalone ./apps/vpnmanager
 COPY --from=vpnmanager-builder --chown=nextjs:nodejs /workspace/apps/vpnmanager/.next/static ./apps/vpnmanager/.next/static
-
+COPY --from=vpnmanager-builder --chown=nextjs:nodejs /workspace/apps/vpnmanager/scripts ./apps/vpnmanager/scripts
+COPY --from=vpnmanager-builder --chown=nextjs:nodejs /workspace/apps/vpnmanager/app.json ./app.json
 USER nextjs
 
+ENV SECRET_TOKEN=${SECRET_TOKEN}
```
Comment (kilemensi) [inline Dockerfile:608]:
```suggestion
```

`ENV` are stored in the image after build (`ARG` disappear after build)... don't store anything sensitive in `ENV` when building images.

File: apps/vpnmanager/scripts/processGsheet.js [hunk 3]
```diff
@@ -0,0 +1,12 @@
+async function main() {
+  const { NEXT_PUBLIC_APP_URL, SECRET_TOKEN } = process.env;
```
Comment (kilemensi) [inline apps/vpnmanager/scripts/processGsheet.js:2]: May be it works (as this script is probably called outside of Next.js), but when it comes to Next.js apps it's best not to [destructure](https://nextjs.org/docs/app/api-reference/next-config-js/env) process.env:
>Next.js will replace process.env.customKey with 'my-value' at build time. Trying to destructure process.env variables won't work due to the nature of webpack [DefinePlugin](https://webpack.js.org/plugins/define-plugin/).

File: apps/vpnmanager/src/pages/api/processGsheet.ts [hunk 4]
```diff
@@ -0,0 +1,15 @@
+import { NextApiResponse, NextApiRequest } from "next";
+import { processNewUsers } from "@/vpnmanager/lib/processUsers";
+
+export async function handler(req: NextApiRequest, res: NextApiResponse) {
+  try {
+    const key: string = req.headers["x-api-key"] as string;
+    const SECRET_TOKEN = process.env.SECRET_TOKEN;
+    if (!key || key !== SECRET_TOKEN) {
```
Comment (kilemensi) [inline apps/vpnmanager/src/pages/api/processGsheet.ts:8]:
```suggestion
    if (!(key && key !== SECRET_TOKEN)) {
```

File: apps/vpnmanager/scripts/processGsheet.js [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+  const headers = {
+    "x-api-key": SECRET_TOKEN ?? "",
+  };
+  const res = await fetch(`${NEXT_PUBLIC_APP_URL}/api/processGsheet`, {
+    headers,
+  });
+  return res.json();
+}
+
+main().then(console.log);
```
Comment (kilemensi) [inline apps/vpnmanager/scripts/processGsheet.js:12]: 1. I'm assuming this is not logging anything sensitive?
2. Why not use await?

File: .github/workflows/vpnmanager-deploy-dev.yml (hunk 1 above)
Comment (koechkevin) [reply .github/workflows/vpnmanager-deploy-dev.yml:60]: API secret key to ensure the script is not abused.

File: Dockerfile [hunk 6]
```diff
@@ -582,6 +582,7 @@ RUN pnpm --filter "./apps/vpnmanager" build
 
 FROM base-runner as vpnmanager-runner
 
+ARG SECRET_TOKEN
```
Comment (kilemensi) [inline Dockerfile:585]:
```suggestion
ARG API_SECRET_KEY
```

File: Dockerfile [hunk 7]
```diff
@@ -599,7 +600,8 @@ COPY --from=vpnmanager-builder --chown=nextjs:nodejs /workspace/apps/vpnmanager/
 # https://nextjs.org/docs/advanced-features/output-file-tracing
 COPY --from=vpnmanager-builder --chown=nextjs:nodejs /workspace/apps/vpnmanager/.next/standalone ./apps/vpnmanager
 COPY --from=vpnmanager-builder --chown=nextjs:nodejs /workspace/apps/vpnmanager/.next/static ./apps/vpnmanager/.next/static
-
+COPY --from=vpnmanager-builder --chown=nextjs:nodejs /workspace/apps/vpnmanager/scripts ./apps/vpnmanager/scripts
+COPY --from=vpnmanager-builder --chown=nextjs:nodejs /workspace/apps/vpnmanager/app.json ./app.json
```
Comment (kilemensi) [inline Dockerfile:604]: This should be in contrib/dokku folder (I'm assuming it's only used by Dokku)

**Review by kilemensi — APPROVED** [review]
🚀 

--

Doesn't turbo complain about new env vars such as `API_SECRET_KEY`?

File: Dockerfile (hunk 7 above)
Comment (koechkevin) [reply Dockerfile:604]: @kilemensi `app.json` should be at the root for the scripts to be executed

Comment (kilemensi) [reply Dockerfile:604]: Location of `app.json` is [configurable](https://dokku.com/docs/advanced-usage/deployment-tasks/?h=app.json+location#changing-the-appjson-location) @koechkevin

**Pull request merged**